### PR TITLE
[FEA] Remove FAISS dependency, inherit other common dependencies from raft

### PR DESCRIPTION
**Pull request by trxcllnt**
Originally this PR was about building FAISS shared libs via CPM, but @rlratzel mentioned cuGraph doesn't need FAISS anymore, and it'd be better if we remove it.

If we need FAISS again in the future, we can add `faiss::faiss` back to `target_link_libraries` without needing extra CPM configuration as it will be available [via raft](https://github.com/rapidsai/raft/pull/345).

Edit: Removed more dependencies that we inherit from raft and/or rmm. Depends on https://github.com/rapidsai/raft/pull/345.

Edit 2: I think the CUDA 11.0 thrust issue will be solved by https://github.com/rapidsai/rapids-cmake/pull/98

### CODE REVIEW

**Comment by BradReesWork** [comment]
@trxcllnt is this really a hotfix for 21.10?

**Comment by ajschmidt8** [comment]
rerun tests

File: cpp/cmake/thirdparty/get_faiss.cmake [hunk 1]
```diff
@@ -24,22 +24,34 @@ function(find_and_configure_faiss)
...
-        GLOBAL_TARGETS  faiss FAISS::FAISS
+        GLOBAL_TARGETS   faiss FAISS::FAISS
+        BUILD_EXPORT_SET cugraph-exports
         CPM_ARGS
           GIT_REPOSITORY  https://github.com/facebookresearch/faiss.git
           GIT_TAG         ${PKG_PINNED_TAG}
           OPTIONS
             "FAISS_ENABLE_PYTHON OFF"
-            "BUILD_SHARED_LIBS OFF"
+            "BUILD_SHARED_LIBS ${BUILD_SHARED_LIBS}"
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/get_faiss.cmake:40]: No need the variable will get propagated automatically

File: cpp/cmake/thirdparty/get_faiss.cmake [hunk 2]
```diff
@@ -24,22 +24,34 @@ function(find_and_configure_faiss)
...
             "FAISS_ENABLE_GPU ON"
             "BUILD_TESTING OFF"
             "CMAKE_MESSAGE_LOG_LEVEL VERBOSE"
     )
 
     if(FAISS_ADDED)
-      set(FAISS_GPU_HEADERS ${FAISS_SOURCE_DIR} PARENT_SCOPE)
+        target_include_directories(faiss INTERFACE $<BUILD_INTERFACE:${FAISS_SOURCE_DIR}>)
+        install(TARGETS faiss EXPORT faiss-exports)
+        rapids_export(BUILD faiss
```
Comment (robertmaynard) [inline cpp/cmake/thirdparty/get_faiss.cmake:50]: Since FAISS is a private dependency of cugraph why do we need to install / export this at all?

**Comment by BradReesWork** [comment]
@trxcllnt by pushing dependencies to RAFT, how will cuGraph know what it is dependent on and what libraries are being brought into the final package?

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

File: build.sh [hunk 3]
```diff
@@ -11,6 +11,20 @@
...
+#         export PATH="/usr/local/gcc9/bin:$PATH"
+#     fi
+#     if [ -d /usr/local/binutils/bin ]; then
+#         export PATH="/usr/local/binutils/bin:$PATH"
+#     fi
+#     echo -e "build.sh PATH:\n$PATH"
+#     echo -e "build.sh ld:\n$(which -a ld)"
+#     echo -e "build.sh env:\n$(env)"
+# fi
+
```
Comment (rlratzel) [inline build.sh:27]: Can we remove this commented-out block, or is this something that might still be needed?

File: cpp/CMakeLists.txt [hunk 4]
```diff
@@ -444,12 +432,9 @@ cuGraph library is a collection of GPU accelerated graph algorithms that process
 
 ]=])
 
-set(code_string
-[=[
-thrust_create_target(cugraph::Thrust FROM_OPTIONS)
-]=])
+set(code_string "")
```
Comment (rlratzel) [inline cpp/CMakeLists.txt:435]: is it better to do this, or just remove it altogether along with the `FINAL_CODE_BLOCK` args to the `rapids_export` calls below?

File: cpp/cmake/thirdparty/get_raft.cmake [hunk 5]
```diff
@@ -42,6 +42,8 @@ set(CUGRAPH_BRANCH_VERSION_raft "${CUGRAPH_VERSION_MAJOR}.${CUGRAPH_VERSION_MINO
 # To use a different RAFT locally, set the CMake variable
 # RPM_raft_SOURCE=/path/to/local/raft
 find_and_configure_raft(VERSION    ${CUGRAPH_MIN_VERSION_raft}
-                        FORK       rapidsai
-                        PINNED_TAG branch-${CUGRAPH_BRANCH_VERSION_raft}
+                        # FORK       rapidsai
+                        # PINNED_TAG branch-${CUGRAPH_BRANCH_VERSION_raft}
+                        FORK       trxcllnt
+                        PINNED_TAG fix/build-shared-faiss
```
Comment (rlratzel) [inline cpp/cmake/thirdparty/get_raft.cmake:48]: I'm assuming this is just for debugging the CI failures?

**Review by rlratzel — COMMENTED** [review]
Thank you for also remembering to update `build.sh`.

File: cpp/cmake/thirdparty/get_raft.cmake (hunk 5 above)
Comment (trxcllnt) [reply cpp/cmake/thirdparty/get_raft.cmake:48]: Without this, CI in this PR would never go green :-)

File: build.sh (hunk 3 above)
Comment (trxcllnt) [reply build.sh:27]: yes this isn't needed anymore. I'll remove it.

File: cpp/CMakeLists.txt (hunk 4 above)
Comment (trxcllnt) [reply cpp/CMakeLists.txt]: Removed!

File: cpp/CMakeLists.txt [hunk 6]
```diff
@@ -122,37 +124,16 @@ if(OpenMP_FOUND)
...
 
-include(cmake/thirdparty/get_thrust.cmake)
-include(cmake/thirdparty/get_faiss.cmake)
-include(cmake/thirdparty/get_nccl.cmake)
-include(cmake/thirdparty/get_rmm.cmake)
-
 include(cmake/thirdparty/get_raft.cmake)
-
-include(cmake/thirdparty/get_cuco.cmake)
+include(cmake/thirdparty/get_nccl.cmake)
```
[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/CMakeLists.txt:149]: @ChuckHastings, isnt cugraph is still using the RAFT comms?

[the remark this answers is not in the document]
Comment (ChuckHastings) [reply cpp/CMakeLists.txt:149]: Yes, cugraph uses the RAFT comms implementation.

[the remark this answers is not in the document]
Comment (cjnolet) [reply cpp/CMakeLists.txt:149]: Ah, I misunderstood the original statement- you were saying cugraph should be getting nccl transitively through raft.

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

**Comment by rlratzel** [comment]
rerun tests

Rerunning to get the latest updates to fix style checker version check and CMakeLists.txt changes that resolve a build failure.

File: ci/gpu/build.sh [hunk 7]
```diff
@@ -91,6 +91,20 @@ if [[ -z "$PROJECT_FLASH" || "$PROJECT_FLASH" == "0" ]]; then
     gpuci_logger "Build from source"
     $WORKSPACE/build.sh -v clean libcugraph pylibcugraph cugraph
 else
+<<<<<<< HEAD
```
Comment (rlratzel) [inline ci/gpu/build.sh:94]: oops.

**Comment by trxcllnt** [comment]
@robertmaynard and I discussed offline that we need to add a `FAISS` component to raft before we can merge this, as we don't want CMake to download and build faiss when cuGraph doesn't use it.

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by cjnolet** [comment]
We recently merged the PR to raft that establishes multiple output targets- specifically, there is now a libraft-nn package which depends kn FAISS (and which cugraph no longer depends on directly). Does this help at all alleviate the FAISS issue on cugraph?

**Comment by trxcllnt** [comment]
rerun tests

**Comment by cjnolet** [comment]
@trxcllnt, not sure if it helps but I opened https://github.com/rapidsai/cugraph/pull/2009 to fully remove the faiss dependency from cugraph and implement @robertmaynard's recent changes to the raft build.

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by trxcllnt** [comment]
This PR has been updated to use https://github.com/rapidsai/raft/pull/428

**Comment by trxcllnt** [comment]
rerun tests

**Comment by cjnolet** [comment]
rerun tests

**Comment by cjnolet** [comment]
rerun tests

**Comment by trxcllnt** [comment]
rerun tests

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**